File: sdk/python/minisandbox/sandbox.py
"""Sandbox 与 Execution 资源对象：生命周期与命令执行。"""
import base64
import time
import urllib.parse

from .errors import RunError
from .files import SandboxFiles
from .portproxy import PortHTTP
from .pty import PTYConnection
from .transport import Transport, sleep

TERMINAL_EXECUTION_STATES = ("Exited", "Failed", "Cancelled", "TimedOut")


class Client:
    """MiniSandbox 生命周期 API 的同步 Python 客户端。"""

    def __init__(self, base_url, poll_interval=0.25, timeout=60):
        self._transport = Transport(base_url, poll_interval=poll_interval, timeout=timeout)

    def create(self, image, ttl_seconds=None, outbound=None, idempotency_key=None):
        """创建 sandbox 并返回资源对象。"""
        body = {"image": image}
        if ttl_seconds is not None:
            body["ttl_seconds"] = ttl_seconds
        if outbound is not None:
            body["network"] = {"outbound": outbound}
        headers = {}
        if idempotency_key:
            headers["Idempotency-Key"] = idempotency_key
        _, sandbox = self._transport.request_json("POST", "/v1/sandboxes", body, headers)
        return Sandbox(self._transport, sandbox["id"])

    def sandbox(self, sandbox_id):
        """用已知 ID 绑定资源对象；不发起请求。"""
        return Sandbox(self._transport, sandbox_id)

    def health(self):
        """查询控制面存活状态。"""
        self._transport.request_json("GET", "/healthz")

    def readiness(self):
        """查询控制面与必要组件的就绪状态；未就绪不视为错误。"""
        try:
            _, value = self._transport.request_json("GET", "/readyz")
        except Exception:
            raise
        return {
            "ready": value.get("status") == "ready",
            "components": [
                {"name": component["name"], "ready": component["status"] == "ready"}
                for component in value.get("components", [])
            ],
        }


class Sandbox:
    """一个 sandbox 资源；本身不缓存状态。"""

    def __init__(self, transport, sandbox_id):
        self._transport = transport
        self.id = sandbox_id

    def _base(self):
        return "/v1/sandboxes/" + urllib.parse.quote(self.id)

    def info(self):
        """查询当前生命周期状态。"""
        _, value = self._transport.request_json("GET", self._base())
        return _map_sandbox(value)

    def wait_running(self, timeout=90):
        """轮询至 Running；Failed 或提前 Terminated 时抛错。"""
        deadline = time.monotonic() + timeout
        while True:
            info = self.info()
            if info["state"] == "Running":
                return info
            if info["state"] in ("Failed", "Terminated"):
                raise RuntimeError(
                    f"sandbox {self.id} entered {info['state']}: {info['reason']}: {info['message']}")
            if time.monotonic() > deadline:
                raise TimeoutError(f"timeout waiting for sandbox {self.id} to become Running")
            sleep(self._transport.poll_interval)

    def capabilities(self):
        """查询当前 sandbox 的功能能力。"""
        _, value = self._transport.request_json("GET", self._base() + "/capabilities")
        return {"files": value["files"], "pty": value["pty"], "http_port_proxy": value["http_port_proxy"]}

    def wait_ready(self, timeout=90):
        """等待 Running 并确认能力可用。"""
        info = self.wait_running(timeout)
        deadline = time.monotonic() + timeout
        while True:
            try:
                return info, self.capabilities()
            except Exception:
                pass
            if time.monotonic() > deadline:
                raise TimeoutError(f"timeout waiting for sandbox {self.id} capabilities")
            sleep(self._transport.poll_interval)

    def renew(self, expires_at):
        """延长租约到绝对时间。"""
        _, value = self._transport.request_json(
            "POST", self._base() + "/renew", {"expires_at": expires_at.isoformat()})
        return _map_sandbox(value)

    def delete(self):
        """提交删除意图并立即返回。"""
        self._transport.request_json("DELETE", self._base())

    def delete_and_wait(self, timeout=90):
        """删除并等待 Terminated。"""
        self.delete()
        deadline = time.monotonic() + timeout
        while True:
            info = self.info()
            if info["state"] == "Terminated":
                return info
            if info["state"] == "Failed":
                raise RuntimeError(f"sandbox {self.id} failed during deletion: {info['reason']}")
            if time.monotonic() > deadline:
                raise TimeoutError(f"timeout waiting for sandbox {self.id} to terminate")
            sleep(self._transport.poll_interval)

    def start_execution(self, argv=None, shell=None, cwd=None, env=None, timeout_seconds=None):
        """启动后台 execution 并返回资源对象。"""
        body = {"background": True}
        if argv is not None:
            body["argv"] = argv
        if shell is not None:
            body["shell"] = shell
        if cwd is not None:
            body["cwd"] = cwd
        if env is not None:
            body["env"] = env
        if timeout_seconds is not None:
            body["timeout_seconds"] = timeout_seconds
        _, value = self._transport.request_json("POST", self._base() + "/executions", body)
        return Execution(self._transport, self.id, value["execution_id"])

    def run(self, argv=None, shell=None, cwd=None, env=None, timeout_seconds=None, wait_timeout=90):
        """一次调用完成执行并收集输出；非成功终态抛 RunError。"""
        execution = self.start_execution(argv=argv, shell=shell, cwd=cwd, env=env,
                                         timeout_seconds=timeout_seconds)
        info = execution.wait(wait_timeout)
        stdout, stderr = execution.collect_logs(wait_timeout)
        terminal = info.get("terminal_event") or {}
        result = {
            "execution_id": execution.id,
            "state": info["state"],
            "exit_code": terminal.get("exit_code", -1),
            "stdout": stdout,
            "stderr": stderr,
            "duration_ms": terminal.get("duration_ms", 0),
            "output_truncated": terminal.get("output_truncated", False),
        }
        if info["state"] != "Exited" or result["exit_code"] != 0:
            raise RunError(result, f"execution {execution.id} {info['state']} "
                                   f"(exit {result['exit_code']})")
        return result

    def files(self):
        """返回 workspace 文件管理对象。"""
        return SandboxFiles(self._transport, self.id)

    def open_pty(self, argv, cols=80, rows=24, cwd=None, env=None, timeout_seconds=None):
        """打开一个交互式 PTY 会话。"""
        return PTYConnection.open(self._transport, self.id, argv, cols, rows,
                                  cwd=cwd, env=env, timeout_seconds=timeout_seconds)

    def port_http(self):
        """返回 sandbox loopback HTTP 代理对象。"""
        return PortHTTP(self._transport, self.id)


class Execution:
    """sandbox 中的一个后台执行。"""

    def __init__(self, transport, sandbox_id, execution_id):
        self._transport = transport
        self._sandbox_id = sandbox_id
        self.id = execution_id

    def _base(self):
        return ("/v1/sandboxes/" + urllib.parse.quote(self._sandbox_id) +
                "/executions/" + urllib.parse.quote(self.id))

    def info(self):
        """查询当前状态。"""
        _, value = self._transport.request_json("GET", self._base())
        return {
            "execution_id": value["execution_id"],
            "state": value["state"],
            "terminal_event": value.get("terminal_event"),
        }

    def wait(self, timeout=90):
        """等待任一合法终态。"""
        deadline = time.monotonic() + timeout
        while True:
            info = self.info()
            if info["state"] in TERMINAL_EXECUTION_STATES:
                return info
            if time.monotonic() > deadline:
                raise TimeoutError(f"timeout waiting for execution {self.id} terminal state")
            sleep(self._transport.poll_interval)

    def cancel_and_wait(self, timeout=90):
        """取消并等待终态。"""
        self._transport.request_json("DELETE", self._base())
        return self.wait(timeout)

    def log_page(self, cursor):
        """读取一页日志。"""
        _, value = self._transport.request_json("GET", f"{self._base()}/logs?cursor={cursor}")
        return value

    def collect_logs(self, timeout=90):
        """读取完整日志并解码 stdout/stderr。"""
        deadline = time.monotonic() + timeout
        cursor = 0
        stdout = bytearray()
        stderr = bytearray()
        while True:
            page = self.log_page(cursor)
            for event in page.get("events", []):
                if event["type"] in ("stdout", "stderr"):
                    data = base64.b64decode(event.get("data_base64", ""))
                    if event["type"] == "stdout":
                        stdout.extend(data)
                    else:
                        stderr.extend(data)
            cursor = page["next_cursor"]
            if page.get("complete"):
                return bytes(stdout), bytes(stderr)
            if time.monotonic() > deadline:
                raise TimeoutError(f"timeout reading execution {self.id} logs")
            sleep(self._transport.poll_interval)


def _map_sandbox(value):
    return {
        "id": value["id"],
        "state": value["state"],
        "reason": value["reason"],
        "message": value["message"],
        "image": value["image"],
        "expires_at": value["expires_at"],
        "created_at": value["created_at"],
        "updated_at": value["updated_at"],
    }

File: sdk/typescript/src/sandbox.ts
/**
 * Sandbox 资源对象：生命周期与命令执行。
 */
import { ResponseError } from "./errors.js";
import { SandboxFiles } from "./files.js";
import { PTYConnection } from "./pty.js";
import { PortHTTP } from "./portproxy.js";
import { sandboxBasePath, sleep, Transport } from "./transport.js";
import {
  Capabilities,
  CreateSandboxRequest,
  PTYStartRequest,
  ExecutionEvent,
  ExecutionInfo,
  ExecutionLogPage,
  ExecuteRequest,
  isTerminalExecutionState,
  RunResult,
  SandboxInfo,
} from "./wire.js";

interface WireSandbox {
  id: string;
  state: SandboxInfo["state"];
  reason: string;
  message: string;
  image: string;
  expires_at: string;
  created_at: string;
  updated_at: string;
}

interface WireExecutionDescriptor {
  execution_id: string;
  state: ExecutionInfo["state"];
}

interface WireExecutionStatus {
  execution_id: string;
  state: ExecutionInfo["state"];
  terminal_event?: WireExecutionEvent;
}

interface WireExecutionEvent {
  execution_id: string;
  sequence: number;
  timestamp: string;
  type: string;
  data_base64?: string;
  exit_code?: number;
  duration_ms?: number;
  output_truncated?: boolean;
  error_code?: string;
  message?: string;
}

interface WireLogPage {
  events: WireExecutionEvent[];
  next_cursor: number;
  complete: boolean;
}

/** Client 是 MiniSandbox 生命周期 API 的 TypeScript 客户端。 */
export class Client {
  private readonly transport: Transport;

  constructor(baseUrl: string, options: { fetchImpl?: typeof fetch; pollIntervalMs?: number } = {}) {
    this.transport = new Transport({
      baseUrl,
      fetchImpl: options.fetchImpl,
      pollIntervalMs: options.pollIntervalMs,
    });
  }

  /** 创建 sandbox 并返回资源对象；可选幂等 key 支持安全重放。 */
  async create(
    request: CreateSandboxRequest,
    options: { idempotencyKey?: string } = {},
  ): Promise<Sandbox> {
    const headers: Record<string, string> = {};
    if (options.idempotencyKey) {
      headers["Idempotency-Key"] = options.idempotencyKey;
    }
    const wire: Record<string, unknown> = { image: request.image };
    if (request.ttlSeconds !== undefined) {
      wire.ttl_seconds = request.ttlSeconds;
    }
    if (request.network) {
      wire.network = { outbound: request.network.outbound };
    }
    const { value } = await this.transport.requestJSON<WireSandbox>(
      "POST",
      "/v1/sandboxes",
      wire,
      { headers },
    );
    return new Sandbox(this.transport, value.id);
  }

  /** 用已知 ID 绑定资源对象；不发起请求。 */
  sandbox(sandboxId: string): Sandbox {
    return new Sandbox(this.transport, sandboxId);
  }

  /** 查询控制面存活状态。 */
  async health(): Promise<void> {
    await this.transport.requestJSON<unknown>("GET", "/healthz");
  }

  /** 查询控制面与必要组件的就绪状态；未就绪不视为错误。 */
  async readiness(): Promise<{ ready: boolean; components: Array<{ name: string; ready: boolean }> }> {
    let value: { status: string; components: Array<{ name: string; status: string }> };
    try {
      const decoded = await this.transport.requestJSON<typeof value>("GET", "/readyz");
      value = decoded.value;
    } catch (error) {
      if (error instanceof ResponseError && error.status === 503) {
        return { ready: false, components: [] };
      }
      throw error;
    }
    return {
      ready: value.status === "ready",
      components: value.components.map((component) => ({
        name: component.name,
        ready: component.status === "ready",
      })),
    };
  }
}

/** Sandbox 表示一个 sandbox 资源；本身不缓存状态。 */
export class Sandbox {
  constructor(
    private readonly transport: Transport,
    readonly id: string,
  ) {}

  /** 查询当前生命周期状态。 */
  async info(): Promise<SandboxInfo> {
    const { value } = await this.transport.requestJSON<WireSandbox>(
      "GET",
      sandboxBasePath(this.id),
    );
    return wireToSandboxInfo(value);
  }

  /** 轮询至 Running；Failed 或提前 Terminated 时抛错。 */
  async waitRunning(timeoutMs = 90_000): Promise<SandboxInfo> {
    const deadline = Date.now() + timeoutMs;
    for (;;) {
      const info = await this.info();
      if (info.state === "Running") {
        return info;
      }
      if (info.state === "Failed" || info.state === "Terminated") {
        throw new Error(`sandbox ${this.id} entered ${info.state}: ${info.reason}: ${info.message}`);
      }
      if (Date.now() > deadline) {
        throw new Error(`timeout waiting for sandbox ${this.id} to become Running`);
      }
      await sleep(this.transport.pollIntervalMs);
    }
  }

  /** 延长租约到绝对时间。 */
  async renew(expiresAt: Date): Promise<SandboxInfo> {
    const { value } = await this.transport.requestJSON<WireSandbox>(
      "POST",
      `${sandboxBasePath(this.id)}/renew`,
      { expires_at: expiresAt.toISOString() },
    );
    return wireToSandboxInfo(value);
  }

  /** 提交删除意图并立即返回。 */
  async delete(): Promise<void> {
    await this.transport.requestJSON<unknown>("DELETE", sandboxBasePath(this.id));
  }

  /** 删除并等待 Terminated。 */
  async deleteAndWait(timeoutMs = 90_000): Promise<SandboxInfo> {
    await this.delete();
    const deadline = Date.now() + timeoutMs;
    for (;;) {
      const info = await this.info();
      if (info.state === "Terminated") {
        return info;
      }
      if (info.state === "Failed") {
        throw new Error(`sandbox ${this.id} failed during deletion: ${info.reason}`);
      }
      if (Date.now() > deadline) {
        throw new Error(`timeout waiting for sandbox ${this.id} to terminate`);
      }
      await sleep(this.transport.pollIntervalMs);
    }
  }

  /** 查询当前 sandbox 的功能能力。 */
  async capabilities(): Promise<Capabilities> {
    const { value } = await this.transport.requestJSON<Capabilities>(
      "GET",
      `${sandboxBasePath(this.id)}/capabilities`,
    );
    return value;
  }

  /** 等待 Running 并确认能力可用，返回二者结果。 */
  async waitReady(timeoutMs = 90_000): Promise<{ info: SandboxInfo; capabilities: Capabilities }> {
    const info = await this.waitRunning(timeoutMs);
    const deadline = Date.now() + timeoutMs;
    for (;;) {
      try {
        const capabilities = await this.capabilities();
        return { info, capabilities };
      } catch (error) {
        if (!(error instanceof ResponseError) || !error.retryable) {
          throw error;
        }
      }
      if (Date.now() > deadline) {
        throw new Error(`timeout waiting for sandbox ${this.id} capabilities`);
      }
      await sleep(this.transport.pollIntervalMs);
    }
  }

  /** 返回 workspace 文件管理对象。 */
  files(): SandboxFiles {
    return new SandboxFiles(this.transport, this.id);
  }

  /** 打开一个交互式 PTY 会话。 */
  async openPTY(request: PTYStartRequest): Promise<PTYConnection> {
    return PTYConnection.open(this.transport, this.id, request);
  }

  /** 返回 sandbox loopback HTTP 代理对象。 */
  portHTTP(): PortHTTP {
    return new PortHTTP(this.transport, this.id);
  }

  /** 启动后台 execution 并返回资源对象。 */
  async startExecution(request: ExecuteRequest): Promise<Execution> {
    const { value } = await this.transport.requestJSON<WireExecutionDescriptor>(
      "POST",
      `${sandboxBasePath(this.id)}/executions`,
      toWireExecuteRequest(request, true),
    );
    return new Execution(this.transport, this.id, value.execution_id);
  }

  /** 一次调用完成执行并收集输出；非零退出等终态同时返回结果与错误。 */
  async run(request: ExecuteRequest): Promise<RunResult> {
    const execution = await this.startExecution(request);
    const info = await execution.wait();
    const logs = await execution.collectLogs();
    const result: RunResult = {
      executionId: execution.id,
      state: info.state,
      exitCode: info.terminalEvent?.exitCode ?? -1,
      stdout: logs.stdout,
      stderr: logs.stderr,
      durationMs: info.terminalEvent?.durationMs ?? 0,
      outputTruncated: info.terminalEvent?.outputTruncated ?? false,
    };
    if (info.state !== "Exited" || result.exitCode !== 0) {
      const detail = `execution ${execution.id} ${info.state}` +
        (info.terminalEvent?.exitCode !== undefined ? ` (exit ${info.terminalEvent.exitCode})` : "");
      throw new RunError(result, detail);
    }
    return result;
  }
}

/** RunError 携带已收集输出的终态错误。 */
export class RunError extends Error {
  constructor(readonly result: RunResult, detail: string) {
    super(detail);
    this.name = "MiniSandboxRunError";
  }
}

/** Execution 表示 sandbox 中的一个后台执行。 */
export class Execution {
  constructor(
    private readonly transport: Transport,
    private readonly sandboxId: string,
    readonly id: string,
  ) {}

  /** 查询当前状态。 */
  async info(): Promise<ExecutionInfo> {
    const { value } = await this.transport.requestJSON<WireExecutionStatus>(
      "GET",
      `${sandboxBasePath(this.sandboxId)}/executions/${encodeURIComponent(this.id)}`,
    );
    return {
      executionId: value.execution_id,
      state: value.state,
      terminalEvent:
        value.terminal_event === undefined ? undefined : wireToExecutionEvent(value.terminal_event),
    };
  }

  /** 等待任一合法终态。 */
  async wait(timeoutMs = 90_000): Promise<ExecutionInfo> {
    const deadline = Date.now() + timeoutMs;
    for (;;) {
      const info = await this.info();
      if (isTerminalExecutionState(info.state)) {
        return info;
      }
      if (Date.now() > deadline) {
        throw new Error(`timeout waiting for execution ${this.id} terminal state`);
      }
      await sleep(this.transport.pollIntervalMs);
    }
  }

  /** 取消并等待终态。 */
  async cancelAndWait(timeoutMs = 90_000): Promise<ExecutionInfo> {
    await this.transport.requestJSON<unknown>(
      "DELETE",
      `${sandboxBasePath(this.sandboxId)}/executions/${encodeURIComponent(this.id)}`,
    );
    return this.wait(timeoutMs);
  }

  /** 读取一页日志。 */
  async logPage(cursor: number): Promise<ExecutionLogPage> {
    const { value } = await this.transport.requestJSON<WireLogPage>(
      "GET",
      `${sandboxBasePath(this.sandboxId)}/executions/${encodeURIComponent(this.id)}/logs?cursor=${cursor}`,
    );
    return {
      events: value.events.map(wireToExecutionEvent),
      nextCursor: value.next_cursor,
      complete: value.complete,
    };
  }

  /** 读取完整日志并解码 stdout/stderr。 */
  async collectLogs(timeoutMs = 90_000): Promise<{ stdout: Uint8Array; stderr: Uint8Array }> {
    const deadline = Date.now() + timeoutMs;
    let cursor = 0;
    const stdout: number[] = [];
    const stderr: number[] = [];
    for (;;) {
      const page = await this.logPage(cursor);
      for (const event of page.events) {
        if (event.type === "stdout" || event.type === "stderr") {
          const bytes = Buffer.from(event.dataBase64 ?? "", "base64");
          const target = event.type === "stdout" ? stdout : stderr;
          for (const byte of bytes) {
            target.push(byte);
          }
        }
      }
      cursor = page.nextCursor;
      if (page.complete) {
        return { stdout: Uint8Array.from(stdout), stderr: Uint8Array.from(stderr) };
      }
      if (Date.now() > deadline) {
        throw new Error(`timeout reading execution ${this.id} logs`);
      }
      await sleep(this.transport.pollIntervalMs);
    }
  }
}

function wireToExecutionEvent(value: WireExecutionEvent): ExecutionEvent {
  return {
    executionId: value.execution_id,
    sequence: value.sequence,
    timestamp: value.timestamp,
    type: value.type as ExecutionEvent["type"],
    dataBase64: value.data_base64,
    exitCode: value.exit_code,
    durationMs: value.duration_ms,
    outputTruncated: value.output_truncated,
    errorCode: value.error_code,
    message: value.message,
  };
}

function wireToSandboxInfo(value: WireSandbox): SandboxInfo {
  return {
    id: value.id,
    state: value.state,
    reason: value.reason,
    message: value.message,
    image: value.image,
    expiresAt: value.expires_at,
    createdAt: value.created_at,
    updatedAt: value.updated_at,
  };
}

function toWireExecuteRequest(request: ExecuteRequest, background: boolean): Record<string, unknown> {
  const wire: Record<string, unknown> = { background };
  if (request.argv !== undefined) {
    wire.argv = request.argv;
  }
  if (request.shell !== undefined) {
    wire.shell = request.shell;
  }
  if (request.cwd !== undefined) {
    wire.cwd = request.cwd;
  }
  if (request.env !== undefined) {
    wire.env = request.env;
  }
  if (request.timeoutSeconds !== undefined) {
    wire.timeout_seconds = request.timeoutSeconds;
  }
  return wire;
}
